client/rpc-spec-v2/src/chain_head/test_utils.rs

```diff
@@ -0,0 +1,329 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use parking_lot::Mutex;
+use sc_client_api::{
+	execution_extensions::ExecutionExtensions, BlockBackend, BlockImportNotification,
+	BlockchainEvents, CallExecutor, ChildInfo, ExecutorProvider, FinalityNotification,
+	FinalityNotifications, FinalizeSummary, ImportNotifications, KeysIter, PairsIter, StorageData,
+	StorageEventStream, StorageKey, StorageProvider,
+};
+use sc_utils::mpsc::{tracing_unbounded, TracingUnboundedSender};
+use sp_api::{CallApiAt, CallApiAtParams, NumberFor, RuntimeVersion};
+use sp_blockchain::{BlockStatus, CachedHeaderMetadata, HeaderBackend, HeaderMetadata, Info};
+use sp_consensus::BlockOrigin;
+use sp_runtime::{
+	generic::SignedBlock,
+	traits::{Block as BlockT, Header as HeaderT},
+	Justifications,
+};
+use std::sync::Arc;
+use substrate_test_runtime::{Block, Hash, Header};
+
+/// The number of retries to ensure the client subscribed to block events.
+const NUM_RETRIES: u64 = 5;
+/// The number of seconds to sleep between each retry.
+const SLEEP_SECS: u64 = 1;
+
+pub struct ChainHeadMockClient<Client> {
+	client: Arc<Client>,
+	import_sinks: Mutex<Vec<TracingUnboundedSender<BlockImportNotification<Block>>>>,
+	finality_sinks: Mutex<Vec<TracingUnboundedSender<FinalityNotification<Block>>>>,
+}
+
+impl<Client> ChainHeadMockClient<Client> {
+	pub fn new(client: Arc<Client>) -> Self {
+		ChainHeadMockClient {
+			client,
+			import_sinks: Default::default(),
+			finality_sinks: Default::default(),
+		}
+	}
+
+	pub async fn trigger_import_stream(&self, header: Header) {
+		// Ensure the client called the `import_notification_stream`.
+		let mut retries = NUM_RETRIES;
+		while retries > 0 && self.import_sinks.lock().is_empty() {
+			retries -= 1;
+			tokio::time::sleep(tokio::time::Duration::from_secs(SLEEP_SECS)).await;
+		}
+
+		// Build the notification.
+		let (sink, _stream) = tracing_unbounded("test_sink", 100_000);
+		let notification =
+			BlockImportNotification::new(header.hash(), BlockOrigin::Own, header, true, None, sink);
+
+		for sink in self.import_sinks.lock().iter_mut() {
+			sink.unbounded_send(notification.clone()).unwrap();
+		}
+	}
+
+	pub async fn trigger_finality_stream(&self, header: Header) {
+		// Ensure the client called the `finality_notification_stream`.
+		let mut retries = NUM_RETRIES;
+		while retries > 0 && self.finality_sinks.lock().is_empty() {
+			retries -= 1;
+			tokio::time::sleep(tokio::time::Duration::from_secs(SLEEP_SECS)).await;
+		}
```

Comment:
```suggestion
		while self.finality_sinks.lock().is_empty() {
			tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
		}
```

Sleeping/waiting in tests is never a good idea. The problem is when the CI is overloaded, this will make the tests flaky. Here it is fine to just wait forever to have a sink ready. In general you could write your test in such a way that it has some big timeout (10 min for example) to let it fail. 10 minutes should be enough, even for an overloaded CI system to finish your small tests.

Reply:
That makes total sense! I've added a `run_with_timeout` helper function that implies a 10min maximum duration to complete this future, thanks!